### DISCUSSION TITLE
"Git is unavailable" when doing `brew update`

### DISCUSSION THREAD

**Discussion by akwala** [Homebrew/discussions#363] [Linux]
...even though _git_ is in PATH:

```
$ which git
/usr/local/bin/git

$ brew --prefix
/home/myname/.linuxbrew

$ echo "$PATH"
/home/myname//.rbenv/shims:/home/myname//.rbenv/bin:/home/myname/.local/bin:/home/myname/bin:/home/myname/.linuxbrew/bin:/home/myname/.linuxbrew/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin
```

The error:

```
$ rbenv local 2.6.6

$ brew update
HOMEBREW_CORE_GIT_REMOTE set: using https://github.com/Homebrew/linuxbrew-core for Homebrew/core Git remote URL.
Error: Git is unavailable
Error: Failure while executing; `/home/myname/.linuxbrew/bin/brew tap homebrew/core` exited with 1.
Error: Git must be installed and in your PATH!
```

Help, please.

**Comment by gromgit** [comment]
What's the output of `brew config` and `brew doctor`?

**Reply by akwala** [reply]
Looks like the _brew_ install is incomplete. Should I uninstall and try installing _brew_ anew?
```
$ brew config
HOMEBREW_VERSION: >=2.2.0 (shallow or no git repository)
ORIGIN: (none)
HEAD: (none)
Last commit: never
Core tap: N/A
HOMEBREW_PREFIX: /home/myname/.linuxbrew
HOMEBREW_REPOSITORY: /home/myname/.linuxbrew/Homebrew
HOMEBREW_CELLAR: /home/myname/.linuxbrew/Cellar
HOMEBREW_MAKE_JOBS: 4
CPU: quad-core 64-bit dunno
Homebrew Ruby: 2.6.6 => /home/myname/.rbenv/versions/2.6.6/bin/ruby
Clang: 10.0 build (parse error)
Git: N/A
Error: no executable curl was found

$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: The following directories do not exist:
/home/myname/.linuxbrew/etc
/home/myname/.linuxbrew/include
/home/myname/.linuxbrew/lib
/home/myname/.linuxbrew/opt
/home/myname/.linuxbrew/sbin
/home/myname/.linuxbrew/share
/home/myname/.linuxbrew/var/homebrew/linked
/home/myname/.linuxbrew/Cellar

You should create these directories and change their ownership to your account.
  sudo mkdir -p /home/myname/.linuxbrew/etc /home/myname/.linuxbrew/include /home/myname/.linuxbrew/lib /home/myname/.linuxbrew/opt /home/myname/.linuxbrew/sbin /home/myname/.linuxbrew/share /home/myname/.linuxbrew/var/homebrew/linked /home/myname/.linuxbrew/Cellar
  sudo chown -R $(whoami) /home/myname/.linuxbrew/etc /home/myname/.linuxbrew/include /home/myname/.linuxbrew/lib /home/myname/.linuxbrew/opt /home/myname/.linuxbrew/sbin /home/myname/.linuxbrew/share /home/myname/.linuxbrew/var/homebrew/linked /home/myname/.linuxbrew/Cellar

Warning: Git could not be found in your PATH.
Homebrew uses Git for several internal functions, and some formulae use Git
checkouts instead of stable tarballs. You may want to install Git:
  brew install git
HOMEBREW_CORE_GIT_REMOTE set: using https://github.com/Homebrew/linuxbrew-core for Homebrew/core Git remote URL.
Error: Git is unavailable
Error: Failure while executing; `/home/myname/.linuxbrew/bin/brew tap homebrew/core` exited with 1.
```

**Comment by gromgit** [comment]
Try running `brew update-reset`.

**Reply by akwala** [reply]
```
$ brew update-reset
==> Fetching /home/myname/.linuxbrew/Homebrew...
You must: brew install git

==> Resetting /home/myname/.linuxbrew/Homebrew...
You must: brew install git
```

**Reply by akwala** [reply]
Also complained that Ruby wasn't installed but Ruby 2.7.1p83 is. This is why I installed Ruby 2.6.6 locally – the complaint said Ruby 2.6 couldn't be found. And _curl_ is also installed even though `brew config` reports that it's missing.

**Reply by gromgit** [reply]
You have a very...interesting setup. What do `git --version` and `curl --version` output?

**Reply by akwala** [reply]
> You have a very...interesting setup.

Can't take credit for that, it's the web hosting service I use.

```
$ git --version
git version 2.27.0

$ curl --version
curl 7.74.0 (amd64-portbld-freebsd11.4) libcurl/7.74.0 OpenSSL/1.1.1i zlib/1.2.11 libidn2/2.3.0 nghttp2/1.41.0
Release-Date: 2020-12-09
Protocols: dict file ftp ftps gopher http https imap imaps mqtt pop3 pop3s rtsp smtp smtps telnet tftp 
Features: alt-svc AsynchDNS HTTP2 HTTPS-proxy IDN IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets
```

**Reply by gromgit** [reply]
Just realized you have `git` in a non-standard location (`/usr/bin` is expected), so when `brew` sanitizes its `PATH`, it's no longer visible. (I assume the same is true for `curl`.)

You also have a lot of Homebrew directories missing, so your install is actually quite broken.

First run the install again, and fix anything it complains about: `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

Then, if `brew config` still can't find a usable git and/or curl, you'll need to install the Homebrew versions: `HOMEBREW_NO_ENV_FILTERING=1 brew install git curl`.

After this, run `brew config` again and confirm it can find everything it needs, then you can `brew` whatever else you need.

**Reply by akwala** [reply]
> First run the install again, and fix anything it complains about: `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

This doesn't work because the hosting service doesn't allow use of _sudo_, so it's not provided. This is why I installed _brew_ using the [Alternative Installation](https://docs.brew.sh/Homebrew-on-Linux#alternative-installation) instructions which seems to be why those directories are missing – are [these instructions](https://docs.brew.sh/Homebrew-on-Linux#alternative-installation) complete and up to-date?

Anyways, I was experimenting to see if I could use Homebrew in this environment. It appears that this is a boundary case in which that isn't easily doable.

**Reply by jonchang** [reply]
It "should" be "possible" to run Homebrew in a bootstrap environment like this, where you can upload tarballs to the server but can't otherwise install things like `curl` and `git`, but it's not supported by the standard installation by any means. You'll need to:

1. extract the tarballs for both Homebrew/brew and Homebrew/linuxbrew-core into their proper paths
2. place the tarball for portable ruby into the Homebrew cache (`brew --cache`) and install it from there
2. place the bottle tarballs for `git` and `curl` and all their dependencies into the Homebrew cache 
3. install all dependencies for `git` and curl` straight from the cache in topological order
4. install root certificates into `$(brew --prefix)/etc/openssl@1.1`
5. set `HOMEBREW_FORCE_BREWED_CURL` and `HOMEBREW_FORCE_BREWED_GIT`
6. initialize git repositories with `brew update`

This is painful to do and we won't support it (but will accept pull requests to ensure that it continues to work, assuming it doesn't make the code horrible or do other bad things that we don't want), but it's certainly possible.

**Reply by gromgit** [reply]
> This doesn't work because the hosting service doesn't allow use of _sudo_, so it's not provided.

The official installer script doesn't use sudo when installing to your home directory. I just ran it in a fresh Ubuntu VM:
```
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Select the Homebrew installation directory
- Enter your password to install to /home/linuxbrew/.linuxbrew (recommended)
- Press Control-D to install to /home/ubuntu/.linuxbrew
- Press Control-C to cancel installation
- ```

**Comment by akwala** [comment]
@gromgit This is what I get (OS: FreeBSD 12.2-RELEASE-p2):
```
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
/bin/bash: line 168: /usr/bin/sw_vers: No such file or directory
environment: line 85: /usr/bin/sudo: No such file or directory
Need sudo access on macOS (e.g. the user myname to be an Administrator)!
/bin/bash: line 1: /usr/bin/sudo: No such file or directory
```
@jonchang's [point](https://github.com/Homebrew/discussions/discussions/363#discussioncomment-241922) is well taken.

**Reply by gromgit** [reply]
> FreeBSD 12.2-RELEASE-p2

That's almost certainly your real problem: Homebrew only supports macOS and Linux, so:
* the `git` and `curl` binaries you get via Jonathan's instructions won't work
* you'd have to _majorly_ hack Homebrew proper to get it to run correctly
* everything would be built from source, so you risk triggering whatever usage limits your provider has set

Your best bet is to ask on a FreeBSD forum if there are any supported Homebrew equivalents.

**Reply by akwala** [reply]
Thanks, @gromgit. I should've realized that.

**Reply by gromgit** [reply]
No worries. I've just pushed a PR that addresses this: https://github.com/Homebrew/install/pull/511
